### PR TITLE
report still invalid objects

### LINKED ISSUE #773: Deprecation: Do not create Objects with invalid data anymore
Remove logging for invalid object properties and decline object creation

### PR DESCRIPTION
**Pull request by foxriver76**
- addresses #773
- added possibility to send information to controller sentry instance to inform core devs

### CODE REVIEW

**Comment by Apollon77** [comment]
> hopefully after 4 years, most code is fixed ;-)

I would (by far) not sign that. We shoudl be very careful merging this! Sounds more like a "hey forum users please grep your logs for ..." (ideally after restarting ioBroker) ... and they we see whats up ... or we use sentry to create events to get a real overview

**Comment by foxriver76** [comment]
At some point we should do this and waiting another four years, idk.. would check how bad it is during alpha.

**Comment by Apollon77** [comment]
Lets discuss in next core meeting

File: packages/adapter/src/lib/adapter/adapter.ts [hunk 1]
```diff
@@ -11667,11 +11661,8 @@ export class AdapterClass extends EventEmitter {
         try {
             tools.validateGeneralObjectProperties(task, true);
         } catch (e) {
-            // todo: in the future we will not create this object
-            this._logger.warn(`${this.namespaceLog} Object ${task._id} is invalid: ${e.message}`);
-            this._logger.warn(
-                `${this.namespaceLog} This object will not be created in future versions. Please report this to the developer.`
-            );
+            this._logger.error(`${this.namespaceLog} Object ${task._id} is invalid: ${e.message}`);
```
Comment (foxriver76) [inline packages/adapter/src/lib/adapter/adapter.ts:11664]: maybe we can also keep this, as these are the instanceObjects from ioPack and probably not much should occur here and if so it is easy to fix

Comment (Apollon77) [reply packages/adapter/src/lib/adapter/adapter.ts:11872]: Yes, I think thats ok ... when dev defines wrong things, it m,ight be ok to dcrash. lets see how it is on alpha

File: packages/cli/src/lib/setup/setupInstall.ts [hunk 2]
```diff
@@ -1026,12 +1026,9 @@ export class Install {
...
             try {
                 tools.validateGeneralObjectProperties(obj);
-            } catch (err) {
-                // todo: in the future we will not create this object
-                console.warn(`host.${hostname} Object ${obj._id} is invalid: ${err.message}`);
-                console.warn(
-                    `host.${hostname} This object will not be created in future versions. Please report this to the developer.`
-                );
+            } catch (e) {
+                console.warn(`host.${hostname} Object ${obj._id} is invalid: ${e.message}`);
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/cli/src/lib/setup/setupInstall.ts:1024]: The same, lets see .. when the install crashes

File: packages/controller/src/main.ts [hunk 3]
```diff
@@ -3005,6 +3005,21 @@ async function processMessage(msg: ioBroker.SendableMessage): Promise<null | voi
             restart(() => !isStopping && stop(false));
             break;
         }
+
+        case 'sendToSentry': {
+            const message: string = msg.message.message;
```
Comment (Apollon77) [inline packages/controller/src/main.ts:3010]: The message just contains the object id ... is this object-id complete? So also contains adapter and namespace?

File: packages/controller/src/main.ts [hunk 4]
```diff
@@ -3005,6 +3005,21 @@ async function processMessage(msg: ioBroker.SendableMessage): Promise<null | voi
...
+
+            const sentryInstance = pluginHandler.getPluginInstance('sentry');
+
+            if (!sentryInstance) {
+                logger.debug(`${hostLogPrefix} Do not send message "${message}" to Sentry, because it is disabled`);
+                return;
+            }
+
+            // @ts-expect-error Plugin is not well typed and SentryPlugin has no types at all currently
+            sentryInstance.getSentryObject().captureException(message);
```
Comment (Apollon77) [inline packages/controller/src/main.ts:3020]: hm ... is this really bringing that much? Then we would have one event per broken object ID ... which could contain device specif stuff so maaaaany keys ...

Additionally I would love to not get them as exceptions.

I think it makes more sense to have it like https://github.com/iobroker-community-adapters/ioBroker.hmip/blob/master/main.js#L976-L979

message = Deprecation<adaptername>
with "info" as severity

and as data object we could set
* "version" of the adapter (i think makes sense)
* "deprecation": the message that contains the error string

**Review by Apollon77 — CHANGES_REQUESTED** [review]

File: packages/controller/src/main.ts (hunk 3 above)
Comment (foxriver76) [reply packages/controller/src/main.ts:3010]: no just the message without logging namespace, we now send adapter name and version in extra

File: packages/adapter/src/lib/adapter/adapter.ts [hunk 5]
```diff
@@ -11939,6 +11937,47 @@ export class AdapterClass extends EventEmitter {
...
+
+        const obj = {
+            command: 'sendToSentry',
+            message: {
+                extraInfo: {
+                    deprecationMessage,
+                    adapter: this.name,
+                    version: this.version
+                },
+                message: 'Deprecation',
```
Comment (Apollon77) [inline packages/adapter/src/lib/adapter/adapter.ts:11972]: Send please "Deprecation adaptername"

**Review by Apollon77 — APPROVED** [review]

**Pull request merged**